source/_posts/2022-11-08-small-stalk.html

```diff
@@ -0,0 +1,234 @@
+    connection even if there are thousands and thousands of clients. Virtual threads are not available in a stable JDK
+    yet, and I used a preview build of the JDK in order to try them out in small-stalk.
+</p>
+
+<h2>Connection Registry</h2>
+<p class="justify">
+    When a client connects to small-stalk, it is assigned an ID and its socket is stored in a registry. Identifying
+    clients is necessary for some commands like reserve and delete.
+</p>
+
+<h2>Queue Service</h2>
+<p class="justify">
+    The Queue Service is a combination of:
+</p>
+<ul>
+    <li>Data structures that maintain the state of jobs,</li>
+    <li>A mutation queue into which incoming mutations are queued, and</li>
+    <li>A mutation thread which reads mutations off of the queue and executes them.</li>
+</ul>
+
+<p class="justify">
+    I knew that I wanted to use some kind of write-ahead log for persistence. This led me to the idea of the mutation
+    queue.
+</p>
+
+<p class="justify">
+    A mutation is a data representation of a change that must be made to the queue data structure. All mutating commands
+    (i.e. commands that are not simple reads like peek) map to a mutation. There are also other mutations (such as
+    <span class="monospaced">::time-to-run-expired</span>) that are triggered by timers rather than by a connection
+    thread.
+</p>
+
+<p class="justify">
+    All mutating commands made by connection threads are converted into mutations and enqueued into the mutation queue,
+    to be later executed by a single mutation thread. This means that all state changes are serialised. While this may
+    seem like an unnecessary performance bottleneck, it affords several benefits:
+</p>
+<ul>
+    <li>The state manipulation code is greatly simplified, because synchronisation isn’t a concern.</li>
+    <li>
+        Having a log of mutations means that we can serialise mutations, send them elsewhere and then replay them to
+        reproduce the server’s exact state. This is how AOF persistence is implemented, and we could also ship them over
+        the network to a replica server to maintain as a hot standby.
+    </li>
+</ul>
+
+<h2>Timers and Blocking</h2>
+<p class="justify">
+    Since connection threads will need to block until a command is executed, mutations are enqueued along with promises
+    to which the mutation thread can deliver return values. reserve, in particular, needs to block until a ready job is
+    available. To facilitate this, if a reserve mutation is processed and no job is available, the reserve mutation
+    along with its return promise is enqueued into a separate queue for waiting reserves. When a new job is added, the
+    mutation thread checks the waiting reserves queue and fulfils a waiting reserve, if any.
+</p>
+
+<p class="justify">
+    The <span class="monospaced">reserve-with-timeout</span> command offers a timeout on the reserve blocking. To
```

Comment:
Should this be "offers a timeout on reserve-blocking"? Slightly confusing

Reply:
Going to leave out this change, for lack of time.